lib/ecto/query/builder/limit_offset.ex

```diff
@@ -5,39 +5,91 @@ defmodule Ecto.Query.Builder.LimitOffset do
   runtime work.
   """
-  @spec build(:limit | :offset, Macro.t, [Macro.t], Macro.t, Macro.Env.t) :: Macro.t
+  @spec build(:limit | :with_ties | :offset, Macro.t(), [Macro.t()], Macro.t(), Macro.Env.t()) ::
+          Macro.t()
   def build(type, query, binding, expr, env) do
-    {query, binding} = Builder.escape_binding(query, binding, env)
-    {expr, {params, _acc}} = Builder.escape(expr, :integer, {[], %{}}, binding, env)
+    {query, vars} = Builder.escape_binding(query, binding, env)
+    {expr, {params, _acc}} = escape(type, expr, {[], %{}}, vars, env)
     params = Builder.escape_params(params)
+    quoted = build_quoted(type, expr, params, env)
+
+    Builder.apply_query(query, __MODULE__, [type, quoted], env)
+  end
+
+  defp escape(type, expr, params_acc, vars, env) when type in [:limit, :offset] do
+    Builder.escape(expr, :integer, params_acc, vars, env)
+  end
+
+  defp escape(:with_ties, expr, params_acc, _vars, _env) when is_boolean(expr) do
+    {expr, params_acc}
+  end
+
+  defp escape(:with_ties, {:^, _, [expr]}, params_acc, _vars, _env) do
+    {quote(do: Ecto.Query.Builder.LimitOffset.with_ties!(unquote(expr))), params_acc}
+  end
 
-    limoff = quote do: %Ecto.Query.QueryExpr{
-                        expr: unquote(expr),
-                        params: unquote(params),
-                        file: unquote(env.file),
-                        line: unquote(env.line)}
+  defp build_quoted(:limit, expr, params, env) do
+    quote do: %Ecto.Query.LimitExpr{
+            expr: unquote(expr),
+            params: unquote(params),
+            with_ties: false,
+            file: unquote(env.file),
+            line: unquote(env.line)
+          }
+  end
 
-    Builder.apply_query(query, __MODULE__, [type, limoff], env)
+  defp build_quoted(:offset, expr, params, env) do
+    quote do: %Ecto.Query.QueryExpr{
+            expr: unquote(expr),
+            params: unquote(params),
+            file: unquote(env.file),
+            line: unquote(env.line)
+          }
   end
 
+  defp build_quoted(:with_ties, expr, _params, _env), do: expr
+
   @doc """
   The callback applied by `build/4` to build the query.
   """
-  @spec apply(Ecto.Queryable.t, :limit | :offset, term) :: Ecto.Query.t
+  @spec apply(Ecto.Queryable.t(), :limit | :offset, term) :: Ecto.Query.t()
   def apply(%Ecto.Query{} = query, :limit, expr) do
     %{query | limit: expr}
   end
+
+  def apply(%Ecto.Query{limit: limit} = query, :with_ties, expr) do
+    %{query | limit: apply_limit(limit, expr)}
+  end
+
   def apply(%Ecto.Query{} = query, :offset, expr) do
     %{query | offset: expr}
   end
+
   def apply(query, kind, expr) do
     apply(Ecto.Queryable.to_query(query), kind, expr)
   end
+
+  defp apply_limit(nil, _with_ties) do
+    Builder.error! "`with_ties` can only be applied to queries containing a `limit`"
+  end
+
+  defp apply_limit(limit, with_ties) do
+    %{limit | with_ties: with_ties}
```

Comment:
@josevalim Sorry to bother. I was trying to make `with_ties` a macro to keep the query compile time as long as possible.

I keep running into an issue here though because during compile time the limit struct is escaped when it gets here most of the time. And sometimes the module name is aliased, sometimes not.

Do you know if I should give up trying to make it a macro and make it a function instead? Or do you think I should be able to get this to work.

Reply:
You can make it a macro but then you need to use the whole builder infrastructure. It has the logic necessary for escaping and unescaping the query to keep it at compile-time. I am fine with whatever decision. :) We can start simpler as well and make it a macro later.

Reply:
Thanks. I think I got it working now. Requested re-review whenever you have the time